python/ray/actor.py

```diff
@@ -900,3 +904,125 @@ def __ray_checkpoint_restore__(self):
+        'time_elapsed_ms_since_last_checkpoint',
+    ],
+)
+"""A namedtuple that contains information about actor's last checkpoint."""
+
+Checkpoint = namedtuple(
+    'Checkpoint',
+    [
+        # ID of this checkpoint.
+        'checkpoint_id',
+        # The timestamp at which this checkpoint was saved,
+        # represented as milliseconds elapsed since Unix epoch.
+        'timestamp',
+    ],
+)
+"""A namedtuple that represents a checkpoint."""
+
+
+class Checkpointable(six.with_metaclass(ABCMeta)):
+    """An interface that indicates an actor can be checkpointed."""
+
+    @abstractmethod
+    def should_checkpoint(self, checkpoint_context):
+        """Whether this actor needs to be checkpointed.
+
+        This method will be called after every task. You should implement this
+        callback to decide whether this actor needs to be checkpointed at this
+        time, based on the checkpoint context, or any other factors.
+
+        Args:
+            checkpoint_context: A namedtuple that contains info about last
+            checkpoint.
+
+        Returns:
+            A boolean value that indicates whether this actor needs to be
+            checkpointed.
+        """
+        pass
+
+    @abstractmethod
+    def save_checkpoint(self, actor_id, checkpoint_id):
+        """Save a checkpoint to persistent storage.
+
+        If `should_checkpoint` returns true, this method will be called. You
+        should implement this callback to save actor's checkpoint and the given
+        checkpoint id to persistent storage.
+
+        Args:
+            actor_id: Actor’s ID.
+            checkpoint_id: ID of this checkpoint. You should save it together
+                with actor’s checkpoint data. And it will be used by the
+                `load_checkpoint` method.
+        Returns:
+            None.
+        """
+        pass
+
+    @abstractmethod
+    def load_checkpoint(self, actor_id, available_checkpoints):
+        """Load actor's previous checkpoint, and restore actor's state.
+
+        This method will be called when an actor is reconstructed, after
+        actor's constructor.
+        If the actor needs to restore from previous checkpoint, this function
+        should restore actor's state and return the checkpoint ID. Otherwise,
+        it should do nothing and return None.
+
+        Args:
+            actor_id: Actor’s ID.
+            available_checkpoints: A list of `Checkpoint` namedtuples that
+            contains all available checkpoint IDs and their timestamps, sorted
+            by timestamp in descending order. Note, this method must return the
+            ID of one checkpoint in this list, or None. Otherwise, an exception
+            will be raised.
+        Returns:
+            The ID of the checkpoint from which the actor was resumed, or None
+            if the actor should restart from the beginning.
+        """
+        pass
+
+    @abstractmethod
+    def checkpoint_expired(self, checkpoint_id):
```

Comment:
@ujvl @stephanie-wang added a `checkpoint_expired` callback here.

Reply:
think you need to pass in the `actor_id` in here as well if it's used to locate the data (since `save_checkpoint` may use it in that way).